### PR TITLE
Improved config loading allowing more options.

### PR DESCRIPTION
**Pull request by Swiftwork**
+ Added configBasedir and config options.
+ Configs can now be loaded from a specified file path or from ".stylelintrc" and "stylelint.config.js".
+ Config can also be specified as objects either in vscode settings or in the package.json.
- Removed useStylelintConfigOverrides because it became obsolete.

### CODE REVIEW

**Comment by Swiftwork** [comment]
I followed your postcss-sorting plugin and vscode-config-resolver pattern ;)

**Comment by mrmlnc** [comment]
Thanks for your work! 💪 I'll review this PR very soon.

File: package.json [hunk 1]
```diff
@@ -41,10 +41,18 @@
       "type": "object",
       "title": "stylefmt configuration",
       "properties": {
-        "stylefmt.useStylelintConfigOverrides": {
```
Comment (mrmlnc) [inline package.json]: I think we should revert this changes, because this feature is related to issue #8.

File: package.json [hunk 2]
```diff
@@ -75,4 +84,4 @@
     "build": "npm run clean && npm run lint && npm run compile",
     "watch": "npm run clean && npm run lint && npm run compile -- --sourceMap --watch"
   }
-}
+}
```
Comment (mrmlnc) [inline package.json:93]: Please, add `FinalNewLine` here.

File: src/extension.ts [hunk 3]
```diff
@@ -78,6 +66,31 @@ function useStylefmt(document: vscode.TextDocument, range: vscode.Range): Promis
...
+	const options = {
+		packageProp: 'stylelint',
+		configFiles: [
+			'.stylelintrc',
+			'stylelint.config.js'
+		],
+		editorSettings: settings.config
+	};
+
+	return configResolver.scan(document.uri.fsPath, options).then((resolved) => {
```
Comment (mrmlnc) [inline src/extension.ts]: Here we should merge resolved config and config from `stylelint.configOverrides` option if `stylefmt.useStylelintConfigOverrides` option is enabled.

File: src/extension.ts [hunk 4]
```diff
@@ -78,6 +66,31 @@ function useStylefmt(document: vscode.TextDocument, range: vscode.Range): Promis
...
+		return stylefmtProcess(document, range, {
+			configBasedir: settings.configBasedir || vscode.workspace.rootPath,
+			config: resolved.json
+		});
+	}).catch(() => {
+		return stylefmtProcess(document, range);
```
Comment (mrmlnc) [inline src/extension.ts]: I think that we shouldn't  work without config. This is abnormal situation and we need to inform the user. You can replace this line on `showOutput(err.toString());`?

**Review by mrmlnc — CHANGES_REQUESTED** [review]
Very good work, but please, correct specified comments. Thanks 👍 🥇

**Comment by mrmlnc** [comment]
Hello, @Swiftwork, I added a few comments, if you agree with me, then please correct them, if not, then I am ready to discuss them.

File: package.json (hunk 1 above)
Comment (Swiftwork) [reply package.json]: Agreed

File: package.json (hunk 2 above)
Comment (Swiftwork) [reply package.json:93]: Will do

File: src/extension.ts (hunk 3 above)
Comment (Swiftwork) [reply src/extension.ts]: Will look into that but sounds reasonable

File: src/extension.ts (hunk 4 above)
Comment (Swiftwork) [reply src/extension.ts]: But stylefmt supports usage without config which indicates to me that that is the appropriate behaviour. However it is up to you, else I believe devs should disable the plugin for workspaces where it is not needed.

Comment (mrmlnc) [reply src/extension.ts]: Yeap, you win 👍

**Comment by Swiftwork** [comment]
I made a new commit with the discussed changes :)

File: src/extension.ts [hunk 5]
```diff
@@ -70,24 +76,30 @@ function stylefmtProcess(document: vscode.TextDocument, range: vscode.Range, con
...
+	if (settingsFmt.useStylelintConfigOverrides) {
+		configOverrides = settingsLint.configOverrides;
+	}
+
+	return configResolver.scan(document.uri.fsPath, resolveOptions).then((resolved) => {
 		return stylefmtProcess(document, range, {
-			configBasedir: settings.configBasedir || vscode.workspace.rootPath,
-			config: resolved.json
+			configBasedir: settingsFmt.configBasedir || vscode.workspace.rootPath,
+			config: extend(true, {}, resolved.json, { rules: configOverrides || {} })
```
Comment (mrmlnc) [inline src/extension.ts]: Please, use `Object.assign` instead of `extend` module.

**Review by mrmlnc — CHANGES_REQUESTED** [review]
Sorry, but we need to remove `extend` module 🌵

File: src/extension.ts (hunk 5 above)
Comment (Swiftwork) [reply src/extension.ts]: Issue with Object.assign is that it doesn't deep copy, so nested properties will be lost rather than overwritten. We could write our own extender, but I didn't see the benefit in that.

Comment (mrmlnc) [reply src/extension.ts]: Yeap, sorry, i forgot that `stylelint` has nested config.

**Review by mrmlnc — APPROVED** [review]

**Comment by mrmlnc** [comment]
I'll merge your PR today, but letter 👍 Thanks for your work!

**Comment by Swiftwork** [comment]
You are welcome, now we just need to exact mimic this behaviour on your postcss-sorting plugin and solve formatOnSave :)

**Comment by mrmlnc** [comment]
If you want add more features, please create issues or PRs  💃

**Comment by Swiftwork** [comment]
If I was unclear, I was referring to another of your vscode plugins. No more additions are planned for this PR and it is ready to merge from my side.
